backend/src/main/java/com/google/sps/AddParticipantHelper.java

```diff
@@ -0,0 +1,212 @@
+      UsernameService usernameService) {
+    this.clock = clock;
+    this.matchDatastore = matchDatastore;
+    this.participantDatastore = participantDatastore;
+    this.userDatastore = userDatastore;
+    this.usernameService = usernameService;
+  }
+
+  /** Add participant do datastore and try to find match immediately */
+  public void doPost(HttpServletRequest request, HttpServletResponse response) throws IOException {
+    // Retrieve JSON object request
+    JSONObject obj = retrieveRequestBody(request);
+    if (obj == null) {
+      response.sendError(HttpServletResponse.SC_BAD_REQUEST, "Could not read request body");
+      return;
+    }
+    JSONObject formDetails = obj.getJSONObject(REQUEST_FORM_DETAILS);
+
+    // Get new Participant from input parameters
+    Participant newParticipant = getParticipantFromInputs(response, formDetails);
+    if (newParticipant == null) {
+      response.sendError(
+          HttpServletResponse.SC_BAD_REQUEST, "Could not get participant from inputs.");
+      return;
+    }
+
+    // Add User to datastore if opted to save preferences
+    boolean savePreference = formDetails.getBoolean(REQUEST_SAVE_PREFERENCE);
+    if (savePreference) {
+      userDatastore.addUser(getUserFromParticipant(newParticipant));
+    }
+
+    // Find immediate match if possible
+    FindMatchQuery query = new FindMatchQuery(clock, participantDatastore);
+    Match match = query.findMatch(newParticipant);
+
+    if (match != null) {
+      // Match found, add to match datastore, update participant datastore
+      long matchId = matchDatastore.addMatch(match);
+
+      // Update current participant entity with new matchId and null availability
+      Participant secondParticipant =
+          participantDatastore.getParticipantFromUsername(match.getSecondParticipantUsername());
+      participantDatastore.addParticipant(secondParticipant.foundMatch(matchId));
+
+      // Add new participant to datastore with new matchId and null availability
+      participantDatastore.addParticipant(newParticipant.foundMatch(matchId));
+      System.out.println("found match");
+    } else {
+      // Match not found, add participant to datastore
+      participantDatastore.addParticipant(newParticipant);
+      System.out.println("match not found");
+    }
+
+    // Confirm received form input
+    response.setContentType("text/plain;charset=UTF-8");
+    response.getWriter().println("Received form input details and queried!");
+  }
+
+  /** Retrieve JSON body payload and convert to a JSONObject for parsing purposes */
+  private JSONObject retrieveRequestBody(HttpServletRequest request) throws IOException {
+    StringBuilder requestBuffer = new StringBuilder();
+    try {
+      BufferedReader reader = request.getReader();
+      String currentLine;
+      while ((currentLine = reader.readLine()) != null) {
+        requestBuffer.append(currentLine);
+      }
+    } catch (IOException e) {
+      return null;
```

Comment:
Please add a Nullable annotation and a short description in the javadocs for when this would return null.